meetings/working-groups/ref-improvements/REF-2022-11-11.md

```diff
@@ -0,0 +1,134 @@
+ref fields to ref struct
+===
+
+# Terminology
+RFRS: `ref` field to `ref struct`
+
+# TypedReference
+All the discussion on RFRS is tied to how we eventually fully support `TypedReference` (remove it from restricted types list). The reason is that `TypedReference` can support storing anything. It can even store a `ref TypedReference` value which makes it equivalent to a type that has RFRS. 
+
+The language does not necessarily need to support this full complexity though. One simplification is that we only limit the arguments passed to `__makeref` to be those we logically support on `ref struct` fields today. That would allow `ref struct` values, or `ref` to normal `struct`, but not `ref` to `ref struct`. This means IL could construct instances that violate C# lifetime rules but that would be labeled as simply unsupported.
+
+However until we know how RFRS support will occur it's difficult to move forward with `TypedReference` support for fear we'd end up painting ourselves into a corner.
+
+# Lifetime notation 
+When discussing `ref` and the associated lifetime challenges it helps to have an explicit notation syntax to fall back on for discussions. That will be done by using the `'a` notation. The letter following `'` is the lifetime name and it can be applied to any value or `ref` in code. By default named lifetimes have no relationship to each other but one can be created by adding `lifetime` constraints at the method body. 
+
+For example here is how the use of `scoped` is expressed in lifetime notation 
+
+```csharp
+Span<int> Read(scoped ref int i) {
+    ...
+}
+
+'b Span<int> Read('a ref 'b int i) {
+
+}
+```
+
+In this case `scoped` becomes the `'a` lifetime. That enforces the behavior of `scoped` because the lifetime for returns is `'b`. There is no provided relationship between `'a` and `'b` hence no conversion and it cannot be returned. 
+
+The *return only* vs. *containing method* scopes can be explained using the following example:
+
+```csharp
+ref struct RS { }
+Span<int> Create(ref RS i) {
+    ...
+}
+
+'b Span<int> Create('b ref 'a RS i) {
+    where 'b <= 'a
+
+}
+```
+
+By establishing a relationship between `'a` and `'b` it allows for both the value and the `ref` to be return from the method. It also prevents cyclic assignment issues because the relationship does not allow for it from a lifetime (`'b` could be smaller than `'a`).
+
+# Complications around ref fields of ref struct
+The immediate issue this causes is it creates layers of lifetimes in our types. Consider concretely: 
+
+```csharp
+ref struct RS { 
+  ref 'a Span<byte> Field; 
+}
+```
+
+The value of the field has a lifetime that is independent of the containing instance. Typically when we look at a value there are two lifetimes to consider: the safe to escape and ref safe to escape. In the case of `RS` though there is also the lifetime of `Field`. 
```

Comment:
>is independent of the containing instance [](http://example.com/codeflow?start=43&length=41)

Is it worth stating that the lifetime of the field (the `ref` and any `ref` within the value) cannot be shorter than the lifetime of the containing instance though?